torch/_inductor/codegen/cpp.py

```diff
@@ -2585,28 +2587,38 @@ def reduction_combine_vec(
             raise NotImplementedError
 
     def indirect_assert(self, var, lower, upper, mask=None):
-        assert not mask, "do not support mask in indirect_indexing assertion"
         assert isinstance(var, CppCSEVariable)
         assert var.dtype is not None
-        if not var.is_vec:
+        if not (var.is_vec or (mask and mask.is_vec)):
             return super().indirect_assert(var, lower, upper, mask)
+
+        var_dtype = var.dtype
         lower_scalar = lower
         upper_scalar = upper
+        var_scalar = var
+        if not var.is_vec:
+            var = f"{self._get_vec_type(var_dtype)}({var})"
+        if mask and not mask.is_vec:
+            mask = f"{self._get_vec_type(var_dtype)}({mask})"
         if lower:
-            lower = f"{self._get_vec_type(var.dtype)}({lower})"
+            lower = f"{self._get_vec_type(var_dtype)}({lower})"
         if upper:
-            upper = f"{self._get_vec_type(var.dtype)}({upper})"
+            upper = f"{self._get_vec_type(var_dtype)}({upper})"
         if lower and upper:
             cond = f"({lower} <= {var}) & ({var} < {upper})"
-            cond_print = f"{lower_scalar} <= {var} < {upper_scalar}"
+            cond_print = f"{lower_scalar} <= {var_scalar} < {upper_scalar}"
         elif lower:
             cond = f"{lower} <= {var}"
-            cond_print = f"{lower_scalar} <= {var}"
+            cond_print = f"{lower_scalar} <= {var_scalar}"
         else:
             assert upper
             cond = f"{var} < {upper}"
-            cond_print = f"{var} < {upper_scalar}"
-        cond = f"({self._get_mask_type(var.dtype)}({cond})).all_masked()"
+            cond_print = f"{var_scalar} < {upper_scalar}"
+        cond = f"{self._get_mask_type(var_dtype)}({cond})"
```

Comment:
I meant something like:
```python
        lower_scalar = lower
        upper_scalar = upper
        if var.is_vec:
            if lower:
                lower = f"{self._get_vec_type(var.dtype)}({lower})"
            if upper:
                upper = f"{self._get_vec_type(var.dtype)}({upper})"
        if lower and upper:
            cond = f"({lower} <= {var}) & ({var} < {upper})"
            cond_print = f"{lower_scalar} <= {var} < {upper_scalar}"
        elif lower:
            cond = f"{lower} <= {var}"
            cond_print = f"{lower_scalar} <= {var}"
        else:
            assert upper
            cond = f"{var} < {upper}"
            cond_print = f"{var} < {upper_scalar}"
        if not var.is_vec:
            cond = f"{self._get_mask_type(var.dtype)}({cond})"
        if mask and not mask.is_vec:
            mask = f"{self._get_mask_type(var.dtype)}({mask})"
```

Doing it this way, we don't vectorize the whoel comparison unnecessarily, we perform the scalar comparison and then we vectorize the result.

Reply:
Thanks for explaining.